src/compiler/checker.ts

```diff
@@ -4173,7 +4173,23 @@ namespace ts {
                     const isLate = isLateBindableName(name);
                     const isWellKnown = isComputedPropertyName(name) && isWellKnownSymbolSyntactically(name.expression);
                     if (!isLate && !isWellKnown && isComputedNonLiteralName(name)) {
-                        return anyType;
+                        const exprType = checkExpression((name as ComputedPropertyName).expression);
+                        let indexerType: Type;
+                        if (isTypeAssignableToKind(exprType, TypeFlags.NumberLike)) {
+                            indexerType = getIndexTypeOfType(parentType, IndexKind.Number);
+                        }
+                        else if (isTypeAssignableToKind(exprType, TypeFlags.StringLike)) {

```

Comment:
string is valid for number indexing as well.. so symbols aside..
`indexrType = (isTypeAssignableToKind(exprType, TypeFlags.NumberLike) && getIndexTypeOfType(parentType, IndexKind.Number) || getIndexTypeOfType(parentType, IndexKind.String)`